### PR TITLE
pumi: new test API

### PR DESCRIPTION
**Pull request by AcriusWinter**
Update standalone test API. See below comment for test results.

Supersedes #35807 (for one package)

https://spack.readthedocs.io/en/latest/packaging_guide.html#stand-alone-testsUpdate standalone test API. See below comment for test results.

### CODE REVIEW

**Comment by AcriusWinter** [comment]
========================= SUMMARY: pumi-2.2.8-cnxm7rx ==========================
Pumi::test_partition .. PASSED
Pumi::test_refine .. PASSED
============================= 2 passed of 2 parts ==============================
============================== 1 passed of 1 spec ==============================

**Comment by AcriusWinter** [comment]
==> Results for test suite 'awtnb3e4swtveiprhk67gjwbhfvykffx':
==> test specs:
==>   pumi-2.2.8-cnxm7rx PASSED
==>   pumi-2.2.6-6e6ejkq SKIPPED
======================== 1 passed, 1 skipped of 2 specs ========================

File: var/spack/repos/builtin/packages/pumi/package.py [hunk 1]
```diff
@@ -111,25 +111,37 @@ def cmake_args(self):
...
-        self.run_test(
-            mpiexec,
-            mpiopt + exe + options,
-            expected,
-            purpose=description,
-            work_dir=self.prefix.bin,
-        )
+    def test_partition(self):
+        """Testing pumi mesh partitioning"""
+        if self.spec.satisfies("@2.2.6"):
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/pumi/package.py]: This is not a check of the range `@:2.2.6`.

File: var/spack/repos/builtin/packages/pumi/package.py [hunk 2]
```diff
@@ -111,25 +111,37 @@ def cmake_args(self):
...
+        options = [
+            "-n",
+            "2",
+            join_path(self.prefix.bin, "split"),
+            join_path(self.prefix.share.testdata, "pipe.dmg"),
+            join_path(self.prefix.share.testdata, "pipe.smb"),
+            "pipe_2_.smb",
+            "2",
+        ]
+        print(f"{options}")
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/pumi/package.py]: Remove the `print()` statements at lines 128, 130, and 132.

File: var/spack/repos/builtin/packages/pumi/package.py [hunk 3]
```diff
@@ -111,25 +111,37 @@ def cmake_args(self):
...
+        exe = which(self.spec["mpi"].prefix.bin.mpiexec)
+        print(f"{exe}")
+        out = exe(*options, output=str.split, error=str.split)
+        print(f"{out}")
+        assert "mesh pipe_2_.smb written" in out
+
+    def test_refine(self):
+        """Testing pumi uniform mesh refinement"""
+        if self.spec.satisfies("@2.2.6"):
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/pumi/package.py]: Not the correct range check.

**Review by tldahlgren — CHANGES_REQUESTED** [review]
Really close.  Just a couple of simple fixes needed.

**Comment by tldahlgren** [comment]
```
$ spack -v test run pumi
==> Spack test hilvdhtv2cjwfqaz6vighup7rqovg6x5
==> Testing package pumi-2.2.8-l6wk753
==> [2024-07-10-16:42:41.077445] test: test_partition: Testing pumi mesh partitioning
...
MDS: reordering before writing smb files
PASSED: Pumi::test_partition
==> [2024-07-10-16:42:41.406248] test: test_refine: Testing pumi uniform mesh refinement
...
MDS: reordering before writing smb files
PASSED: Pumi::test_refine
==> [2024-07-10-16:42:41.808114] Completed testing
==> [2024-07-10-16:42:41.808307] 
========================= SUMMARY: pumi-2.2.8-l6wk753 ==========================
Pumi::test_partition .. PASSED
Pumi::test_refine .. PASSED
============================= 2 passed of 2 parts ==============================
============================== 1 passed of 1 spec ==============================
```

**Review by tldahlgren — APPROVED** [review]
Works for me.

**Comment by tldahlgren** [comment]
@wspear Does this PR fix #38030 for you?

**Pull request merged**